02_Memory_Management/04_Virtual_Memory_Manager.md

```diff
@@ -1,4 +1,203 @@
+
+```c
+uint64_t convert_x86_64_vm_flags(size_t flags) {
+    uint64_t value = 0;
+    if (flags & VM_FLAG_WRITE)
+        value |= PT_FLAG_WRITE;
+    if (flags & VM_FLAG_USER)
+        value |= PT_FLAG_USER;
+    if ((flags & VM_FLAG_EXEC) == 0)
+        value |= PT_FLAG_NX;
+    return value;
+};
+```
+
+The `PT_xyz` macros are just setting the bits in the page table entry, for specifics see the paging chapter. Notice how we set the NX-bit if `VM_FLAG_EXEC` is not set because of a quirk on x86.
+
+We're going to store these vm objects as a linked list, which is the purpose of the `next` field.
+
+### How Many?
+
+Since a virtual memory manager only handles a single address space, you'll need one per address space you wish to have. This roughly translates to one VMM per running program, since each program should live in it's own address space. Later on when we implement scheduling you'll see how this works. 
+
+The kernel is a special case since it should be in all address spaces, as it always needs to be loaded to manage the underlying hardware of the system.
+
+There are many ways of handling this, one example is to have a special kernel VMM that manages all higher half memory, and have other VMMs only manage the lower memory for their respective program. In this design you have a single higher half VMM (for the kernel), and many lower-half VMMs. Only one lower half VMM is active at a time, the one corresponding to the running program.
+
+### Managing An Address Space
+
+This is where design and reality collide, because our high level VMM needs to program the MMU. The exact details of this vary by platform, but for x86(_64) we have paging! See the previous chapter on how x86 paging works. Each virtual memory manager will need to store the appropriate data to manage the address space it controls: for paging we just need the address of the root table.
+
+```c
+void* vmm_pt_root;
+```
+
+## Allocating Objects
+
+Now we know *what* an VM object it, let's look at how we're going to create them. We're going to create a new function for this, our example function is going to have the following prototype:
+
+```c
+void* vmm_alloc(size_t length, size_t flags, void* arg);
+```
+
+The `length` and `flag` fields are how many bytes we want, and the type of memory we want. The `void* arg` field is an oddity, it's a place to pass in extra data if the allocation needs it. We'll see why this is useful later.
+
+The function will try to find a space in virtual memory big enough to hold `length` bytes, if it can it marks this space as taken, and returns the base address. If you're using immediate-backing (as opposed to demand-backing) this is where you would allocate pages from the pmm, and modify the page tables so this physical memory is available.
+
+An example implementation is below, remember we're using a sorted linked list for storing virtual memory objects.
```

Comment:
i would split if possible the introduction of the vmm alloc in several parts, otherwise we are just giving code ready to copy and paste to the reader. The idea is that the reader should be able to implement it by himself reading our guide, not us giving him the code ready.